contracts/Vesting.sol

```diff
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+/// @title Vesting
+/// @author Eldar Dryga
+/// @notice this contract allows you to distribute tokens to users within a certain time
+
+contract Vesting is Ownable, ReentrancyGuard {
+    using SafeERC20 for IERC20;
+    IERC20 private immutable _token;
+
+    uint32 public periodOfDistributionInDays;
+    uint32 public amountOfDistributionPerUser;
+    uint128 public users;
+    uint256 public maxUsers;
+
+    struct UsersInfo {
+        uint256 userTime;
+        uint256 claimedRevard;
+        uint256 allClaimedRevard;
+        uint256 finalDistributionTime;
+    }
+    mapping(address => UsersInfo) public infoOfUser;
+
+    constructor(address token_) {
+        require(token_ != address(0));
+        _token = IERC20(token_);
+    }
+
+    /// @param _amountOfDistribution total number of tokens to give away
+    /// @param _periodOfDistributionInDays the number of days during which the distribution will take place
+    /// @param _amountOfDistributionPerUser the amount of tokens for distribution for 1 user
+    function startVesting(
+        uint256 _amountOfDistribution,
+        uint16 _periodOfDistributionInDays,
+        uint32 _amountOfDistributionPerUser
+    ) public onlyOwner {
+        _token.safeTransferFrom(
+            msg.sender,
+            address(this),
+            _amountOfDistribution
+        );
+        periodOfDistributionInDays = _periodOfDistributionInDays;
+        amountOfDistributionPerUser = _amountOfDistributionPerUser;
+        maxUsers = _amountOfDistribution / amountOfDistributionPerUser;
+    }
+
+    /// @notice this function can only be called by the owner of the contract
+    /// @dev call this function to start vesting
+
+    function join() public nonReentrant {
```

Comment:
Use nonReentrant when in ur function u have contract call to another one, thats the subject of reentrancy issue
You dont have contract calls, so u can remove nonReentrant here